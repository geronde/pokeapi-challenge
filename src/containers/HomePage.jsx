import React, { useEffect, useState } from "react";
import styled from "styled-components";
import { uniqueId, isEmpty } from "lodash";
import Header from "../components/Header";
import Footer from "../components/Footer";
import AppLoader from "../components/Loader";
import Pagination from "../components/Pagination";
import { paginate, getIdFromUrl } from "./utils";
import { Link } from "react-router-dom";


const Wrapper = styled.div`
  width: 100%;
  min-height: 100vh;
  position: relative;
  display: grid;
  grid-template-rows: 1.25fr 6fr 1fr;
  grid-template-areas: "header header header" "content content content" "footer footer footer";
  background: #e2e1e0;
  section {
    grid-area: content;
    display: flex;
    flex-direction: column;
    width: 55%;
    margin: 30px auto;
    color: #000;
  }
  .pokemons-container {
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
  }
  .pokemon {
    padding: 10px;
    margin: 15px;
    width: 150px;
    height: 100px;
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    background: #fff;
    display: flex;
    font-size: 20px;
    align-items: center;
    justify-content: center;
  }
  .pokemon:hover {
    box-shadow: 0 14px 28px rgba(0,0,0,0.25), 0 10px 10px rgba(0,0,0,0.22);  
  }
  .filter{
      padding: 20px;
      width: 50%;
  }
  .filter select {
      width: 150px;
      height: 30px;
  }
`;


const HomePage = () => {
  const [pokemons, setPokemons] = useState([]);

  const [pagination, setPagination] = useState({totalPokemons: null, pageSize: 20});
  const [currentPage, setCurrentPage] = useState(1);
  const [isFilter, setIsFilter] = useState(false)
  const [types, setTypes] = useState([]);
  

  const fetchPokemons = async (url) => {
    const res = await fetch(
      url
    );
    const response = await res.json();
    setPokemons(response.results);
    setPagination({ totalPokemons: response.count, previous: response.previous,next: response.next})    
  };

  const fetchTypes = async (url) => {
    const res = await fetch(
      url
    );
    const response = await res.json();
    setTypes(response.results)
  };

  useEffect(() => {
    fetchPokemons('https://pokeapi.co/api/v2/pokemon');
    fetchTypes('https://pokeapi.co/api/v2/type')
  // eslint-disable-next-line react-hooks/exhaustive-deps
  },[]);

  const filterByType = async (event) => {
    setIsFilter(true)
    const res = await fetch(
        `${event.target.value}`
      );
      const response = await res.json();
      const pokemons = response.pokemon.map((poke)=>poke.pokemon).slice(detail.startIndex, detail.endIndex)
      console.log('pokemons: ', pokemons);
      setPokemons(pokemons) 
      setPagination({...pagination, totalPokemons: pokemons.length, next: null, previous: null})
  }

  const nextPage = () => {
      const activePage = currentPage + 1;
      setCurrentPage(activePage)
      if(isFilter){
        const list = pokemons.slice(detail.startIndex, detail.endIndex)
        setPokemons(list)
      }else {
        fetchPokemons(pagination.next)
      }   
  }

  const previousPage = () => {
    const activePage = currentPage - 1;
    setCurrentPage(activePage)
    if(isFilter){
      const list = pokemons.slice(detail.startIndex, detail.endIndex)
      setPokemons(list)
    }else {
      fetchPokemons(pagination.previous)
    }
  }

const { totalPokemons, pageSize, next, previous } = pagination;
const detail = paginate(totalPokemons, currentPage, pageSize)

  if(isEmpty(pokemons) && !isFilter) return <AppLoader/>

  return (
    <Wrapper>
      <Header>Pokemon</Header>
      <section>
      <article className="filter">
        <select onChange={filterByType}>
        <option disabled selected value> -- select an option -- </option>
          {types.map((type)=><option value={type.url}>{type.name}</option> )}
        </select>
        </article>
        <article className="pokemons-container">
          {pokemons?.map((pokemon, index) => (
            <Link to={`/pokemon/${getIdFromUrl(pokemon.url)}`}><div className="pokemon" key={uniqueId()}>
              {pokemon.name}
            </div>
            </Link>
          ))}
          {isEmpty(pokemons) && isFilter && <div>No results</div>}
        </article>
        {!isEmpty(pokemons) && <Pagination details={{...detail, next, previous}} nextPage={nextPage} previousPage={previousPage} />}
      </section>
      <Footer/>
    </Wrapper>
  );
};

export default HomePage;
