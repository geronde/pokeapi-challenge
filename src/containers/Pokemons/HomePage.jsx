import React, { useEffect, useState, useContext } from "react";
import styled from "styled-components";
import { uniqueId, isEmpty } from "lodash";
import { Link } from "react-router-dom";


import Header from "../../components/Header";
import Footer from "../../components/Footer";
import AppLoader from "../../components/Loader";
import Pagination from "../../components/Pagination";
import { LocaleContext } from '../LocaleProvider/index'


import { paginate, getIdFromUrl } from "./utils";
import messages from './messages';


const Wrapper = styled.div`
  width: 100%;
  min-height: 100vh;
  position: relative;
  display: grid;
  grid-template-rows: 1.25fr 6fr 1fr;
  grid-template-areas: "header header header" "content content content" "footer footer footer";
  background: #e2e1e0;
  section {
    grid-area: content;
    display: flex;
    flex-direction: column;
    width: 75%;
    margin: 0 auto;
    color: #000;
  }
  .pokemons-container {
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    align-items:center;
    width: 90%;
    margin: 0 auto;
  }
  .listing-container {
    width: 95%;
    margin: 0 auto;
  }
  .pokemon {
    padding: 10px;
    margin: 15px;
    width: 150px;
    height: 100px;
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
    transition: all 0.3s cubic-bezier(.25,.8,.25,1);
    background: #fff;
    display: flex;
    font-size: 20px;
    align-items: center;
    justify-content: center;
  }
  .pokemon:hover {
    box-shadow: 0 14px 28px rgba(0,0,0,0.25), 0 10px 10px rgba(0,0,0,0.22);  
  }
  .filter{
      padding: 20px;
      width: 50%;
  }
  .filter select {
      width: 150px;
      height: 30px;
  }
`;


export const HomePage = () => {
  const [pokemons, setPokemons] = useState([]);
  const [types, setTypes] = useState([]);
  const [currentPage, setCurrentPage] = useState(1)
  const [paginationAPI, setPaginationAPI] = useState({ totalItems: 0, next: null, previous: null, pageSize: 20 });
  const [pagination, setPagination] = useState({})
  const [isFilter, setIsFilter] = useState(false);


  const { translate } = useContext(LocaleContext);
  const defaultSelectValue = translate(messages.selectPlaceholder)
  const [selectedType, setSelectedType] = useState(defaultSelectValue)

  const fetchPokemons = async (url) => {
    const res = await fetch(
      url
    );
    const response = await res.json();
    setPokemons(response.results);
    setPaginationAPI({ ...paginationAPI, totalItems: response.count, previous: response.previous, next: response.next })
  };


  useEffect(() => {
    setPagination(paginate(paginationAPI.totalItems, currentPage, paginationAPI.pageSize))
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentPage, paginationAPI.totalItems])


  const fetchTypes = async (url) => {
    const res = await fetch(
      url
    );
    const response = await res.json();
    setTypes(response.results)
  };


  useEffect(() => {
    fetchPokemons('https://pokeapi.co/api/v2/pokemon');
    fetchTypes('https://pokeapi.co/api/v2/type')
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);



  const nextPage = () => {
    const activePage = currentPage + 1;
    setCurrentPage(activePage)
    noFilterFetch(paginationAPI.next)
  }

  const previousPage = () => {
    const activePage = currentPage - 1;
    setCurrentPage(activePage)
    noFilterFetch(paginationAPI.previous)
  }

  const noFilterFetch = (url) => {
    if (!isFilter) {
      fetchPokemons(url)
    }
  }

  const filterByType = async (event) => {
    const selectedOption = event.target[event.target.selectedIndex]
    setSelectedType(selectedOption.value)
    setIsFilter(true);

    const res = await fetch(selectedOption.value);
    const response = await res.json();
    const pokemons = response.pokemon.map((poke) => poke.pokemon)
    setPokemons(pokemons)
    setCurrentPage(1)
    setPaginationAPI({ ...paginationAPI, totalItems: response.pokemon.length })
  }

  const formatPokemonsByFilter = () => {
    if(!isFilter) return pokemons

    return pokemons.slice(pagination.startIndex, pagination.endIndex)
  }

  if (isEmpty(pokemons) && !isFilter) return <AppLoader />
  return (
    <Wrapper>
      <Header>{translate(messages.appTitle)}</Header>
      <section>
        <article className="filter">
          <select value={selectedType} onChange={filterByType}>
            <option disabled>{defaultSelectValue}</option>
            {types.map((type) => <option name={type.name} key={uniqueId(`${type.name}`)} value={type.url}>{type.name}</option>)}
          </select>
        </article>
        <article className="listing-container">
          <div className="pokemons-container">
            {formatPokemonsByFilter().map((pokemon) => (
              <Link to={`/pokemon/${getIdFromUrl(pokemon.url)}`} key={uniqueId()}><div data-testid="single-pokemon" className="pokemon">
                {pokemon.name}
              </div>
              </Link>
            ))}
          </div>
          {isEmpty(pokemons) && isFilter && <div>{translate(messages.noResult)}</div>}
        </article>
        {!isEmpty(pokemons) && <Pagination details={{ ...paginationAPI, currentPage, isFilter, startIndex: pagination.startIndex, endIndex: pagination.endIndex }} nextPage={nextPage} previousPage={previousPage} />}
      </section>
      <Footer />
    </Wrapper>
  );
};

export default HomePage;