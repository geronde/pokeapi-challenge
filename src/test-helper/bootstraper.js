import faker from 'faker';

export const generateTestPokemons = (n = 15, isType) => {
  const pokemons = [];

  for (let i = 0; i < n; i++) {
    let poke = { name: faker.random.word(), url: `fake-url/${i}/` };
    if (isType) {
      poke = { pokemon: { ...poke } };
    }
    pokemons.push(poke);
  }
  return pokemons;
};
