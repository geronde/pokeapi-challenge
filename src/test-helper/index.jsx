import { render as rtlRender } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

export const render = (ui, { ...renderOptions } = {}) =>{
    function Wrapper({ children }) {
      return <BrowserRouter>{children}</BrowserRouter>
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
  }