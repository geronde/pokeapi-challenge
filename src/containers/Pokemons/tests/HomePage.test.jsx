import React from 'react';
import { act, fireEvent, waitFor, screen, cleanup } from '@testing-library/react'

import { render } from '../../../test-helper';
import { generateTestPokemons } from '../../../test-helper/bootstraper';

import HomePage from '../HomePage';

afterEach(cleanup);

describe('HomePage', () => {
    let pokemonResponse;
    let typeResponse;
    let pokemonsByType
    beforeEach(() => {
        pokemonResponse = jest.fn(() => Promise.resolve({ count: 200, previous: '/digi/78', next: '/pika/30/', results: generateTestPokemons(20) }));

        typeResponse = jest.fn(() => Promise.resolve({ results: generateTestPokemons(30) }));

        pokemonsByType = jest.fn(() => Promise.resolve({ pokemon: generateTestPokemons(40, true) }));
    });

    afterEach(() => {
        jest.resetAllMocks();
        jest.resetAllMocks();
    });

    it('should render a loader when pokemon state is empty', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: jest.fn(() => Promise.resolve({ results: [], count: 0 })) });
        await act(async () => {
            const { asFragment } = render(<HomePage />)
            expect(asFragment()).toMatchSnapshot();
        })
    });

    it('should fetch pokemons and pokemons type on page load', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValue({ json: typeResponse })
        await act(async () => {
            render(<HomePage />)
        })

        expect(window.fetch).toHaveBeenCalledTimes(2);
        expect(screen.getByText("Show all pokemons")).toBeInTheDocument()
        expect(screen.getAllByTestId('single-pokemon')).toHaveLength(20)
    });

    it('should be able to select type and display pokemons by type', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValue({ json: pokemonsByType })

        await act(async () => {
            render(<HomePage />)
        });

        const select = screen.getByRole('combobox');
        fireEvent.change(select, { target: { value: 'fake-url/3/' } })
        await waitFor(() => {
            expect(window.fetch).toHaveBeenCalledTimes(3);
        })
        expect(screen.getAllByTestId('single-pokemon')).toHaveLength(19)
    });

    it('should be able to click on nextPage and reload pokemons', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValue({ json: pokemonResponse })

        await act(async () => {
            render(<HomePage />)
        });

        expect(window.fetch).toHaveBeenCalledTimes(2)

        await act(async () => {
            const nextPage = await screen.getAllByRole('button')
            fireEvent.click(nextPage[1])
        })
        expect(window.fetch).toHaveBeenCalledTimes(3)
        expect(window.fetch).toHaveBeenNthCalledWith(3, "/pika/30/")
        expect(screen.getByText("2 of 10")).toBeInTheDocument()
    });

    it('should be able to click on previous and reload pokemons', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValueOnce({ json: pokemonResponse }).mockReturnValue({ json: pokemonResponse })

        await act(async () => {
            render(<HomePage />)
        });

        expect(window.fetch).toHaveBeenCalledTimes(2)

        //click on next to be able to click on previous
        await act(async () => {
            const nextPage = await screen.getAllByRole('button')
            fireEvent.click(nextPage[1])
        })
        expect(window.fetch).toHaveBeenCalledTimes(3)
        expect(window.fetch).toHaveBeenNthCalledWith(3, "/pika/30/")

        await act(async () => {
            const previousPage = await screen.getAllByRole('button')
            fireEvent.click(previousPage[0])
        })
        expect(window.fetch).toHaveBeenCalledTimes(4)
        expect(window.fetch).toHaveBeenNthCalledWith(4, "/digi/78")
        expect(screen.getByText("1 of 10")).toBeInTheDocument()
    });

    it('should be able to click on nextPage and reload pokemons when using filter', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValue({ json: pokemonsByType })

        await act(async () => {
            render(<HomePage />)
        });

        expect(window.fetch).toHaveBeenCalledTimes(2);

        await act(async () => {
            const select = await screen.getByRole('combobox')
            fireEvent.change(select, {target: {value: 'fake-url/1/'}})
        });
        
        expect(window.fetch).toHaveBeenCalledTimes(3);
        expect(window.fetch).toHaveBeenNthCalledWith(3, "fake-url/1/")

        await act(async () => {
            const nextPage = await screen.getAllByRole('button')
            fireEvent.click(nextPage[1])
        })
        expect(screen.getByText("2 of 2")).toBeInTheDocument()
    });

    it('should be able to click on previous page and reload pokemons when using filter', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValue({ json: pokemonsByType })

        await act(async () => {
            render(<HomePage />)
        });

        expect(window.fetch).toHaveBeenCalledTimes(2);

        await act(async () => {
            const select = screen.getByRole('combobox')
            fireEvent.change(select, { target: { value: 'fake-url/2/' } })
        });
        expect(window.fetch).toHaveBeenCalledTimes(3);

        await act(async () => {
            const nextPage = await screen.getAllByRole('button')
            fireEvent.click(nextPage[1])
        })

        await act(async () => {
            const previousPage = await screen.getAllByRole('button')
            fireEvent.click(previousPage[0])
        })

        expect(window.fetch).toHaveBeenNthCalledWith(3, "fake-url/2/")
        expect(screen.getByText("1 of 2")).toBeInTheDocument()
    });

    it('should display no result when no pokemons are return by type filter', async () => {
        const noResult = jest.fn(() => Promise.resolve({ pokemon: [] }))
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemonResponse }).mockReturnValueOnce({ json: typeResponse }).mockReturnValue({ json: noResult })

        await act(async () => {
            render(<HomePage />)
        });

        expect(window.fetch).toHaveBeenCalledTimes(2);

        await act(async () => {
            const select = screen.getByRole('combobox')
            fireEvent.change(select, { target: { value: 'fake-url/2/' } })
        });
        expect(window.fetch).toHaveBeenCalledTimes(3);
        expect(window.fetch).toHaveBeenNthCalledWith(3, "fake-url/2/")
        expect(screen.getByText('No result')).toBeInTheDocument();
    });
})