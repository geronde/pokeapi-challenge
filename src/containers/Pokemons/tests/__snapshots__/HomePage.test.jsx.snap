// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render a loader when pokemon state is empty 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <div
      aria-busy="true"
      class=""
    >
      <svg
        aria-label="audio-loading"
        height="100"
        stroke="#0E1111"
        viewBox="0 0 44 44"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
          stroke-width="2"
        >
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
